From: sunshineinabox <aqemail@gmail.com>
Date: Mon, 27 Jul 2026 20:28:00 -0700
Subject: [PATCH] ASoC: codecs: aw88166: support changing sample rate and bit
 width

The aw88166 driver only worked with the stream configuration baked
into the loaded acf profile due to the lack of a proper PLL
initialization sequence. Fix by selecting all the necessary PLL
settings based on what was passed to us by the hw_params/set_fmt ops.
This replaces the strange downstream routine that tries two divider
modes in sequence.

Adapted for aw88166 from Val Packett's aw88261 series:
https://lore.kernel.org/all/20260529200507.…-val@packett.cool/
Register layout verified against Awinic's downstream
---
 sound/soc/codecs/aw88166.c | 236 ++++++++++++++++++++++++++++---------
 sound/soc/codecs/aw88166.h | 107 +++++++++++++++++
 2 files changed, 288 insertions(+), 55 deletions(-)
--- a/sound/soc/codecs/aw88166.c
+++ b/sound/soc/codecs/aw88166.c
@@ -14,6 +14,7 @@
 #include <linux/minmax.h>
 #include <linux/regmap.h>
 #include <sound/soc.h>
+#include <sound/pcm_params.h>
 #include "aw88166.h"
 #include "aw88395/aw88395_device.h"
 
@@ -30,6 +31,14 @@
 	unsigned int vcalb_init_val;
 	unsigned int re_init_val;
 	unsigned int dither_st;
+
+	unsigned int sr_value;
+	unsigned int cco_mux_value;
+	unsigned int fs_value;
+	unsigned int bck_value;
+	unsigned int bck_inv_value;
+	unsigned int md_value;
+
 	bool phase_sync;
 };
 
@@ -98,7 +107,7 @@
 	return 0;
 }
 
-static int aw_dev_check_mode1_pll(struct aw_device *aw_dev)
+static int aw_dev_check_pll(struct aw_device *aw_dev)
 {
 	int ret, i;
 
@@ -115,71 +124,48 @@
 	return -EPERM;
 }
 
-static int aw_dev_check_mode2_pll(struct aw_device *aw_dev)
+static int aw88166_dev_configure_syspll(struct aw88166 *aw88166)
 {
-	unsigned int reg_val;
-	int ret, i;
+	struct aw_device *aw_dev = aw88166->aw_pa;
+	int ret;
 
-	ret = regmap_read(aw_dev->regmap, AW88166_PLLCTRL2_REG, &reg_val);
+	/* PLL divider must be used for 8/16/32 kHz modes */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_PLLCTRL2_REG,
+			~AW88166_CCO_MUX_MASK, aw88166->cco_mux_value);
 	if (ret)
 		return ret;
 
-	reg_val &= (~AW88166_CCO_MUX_MASK);
-	if (reg_val == AW88166_CCO_MUX_DIVIDED_VALUE) {
-		dev_dbg(aw_dev->dev, "CCO_MUX is already divider");
-		return -EPERM;
-	}
-
-	/* change mode2 */
-	ret = regmap_update_bits(aw_dev->regmap, AW88166_PLLCTRL2_REG,
-			~AW88166_CCO_MUX_MASK, AW88166_CCO_MUX_DIVIDED_VALUE);
+	/* The word clock (WCK) defines the beginning of a frame */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_I2SCTRL1_REG,
+			~AW88166_I2SSR_MASK, aw88166->sr_value);
 	if (ret)
 		return ret;
 
-	for (i = 0; i < AW88166_DEV_SYSST_CHECK_MAX; i++) {
-		ret = aw_dev_get_iis_status(aw_dev);
-		if (ret) {
-			dev_err(aw_dev->dev, "mode2 iis signal check error");
-			usleep_range(AW88166_2000_US, AW88166_2000_US + 10);
-		} else {
-			break;
-		}
-	}
-
-	/* change mode1 */
-	regmap_update_bits(aw_dev->regmap, AW88166_PLLCTRL2_REG,
-			~AW88166_CCO_MUX_MASK, AW88166_CCO_MUX_BYPASS_VALUE);
-	if (ret == 0) {
-		usleep_range(AW88166_2000_US, AW88166_2000_US + 10);
-		for (i = 0; i < AW88166_DEV_SYSST_CHECK_MAX; i++) {
-			ret = aw_dev_get_iis_status(aw_dev);
-			if (ret) {
-				dev_err(aw_dev->dev, "mode2 switch to mode1, iis signal check error");
-				usleep_range(AW88166_2000_US, AW88166_2000_US + 10);
-			} else {
-				break;
-			}
-		}
-	}
+	/* The bit clock (BCK) defines the length of a frame */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_I2SCTRL1_REG,
+			~AW88166_I2SBCK_MASK, aw88166->bck_value);
+	if (ret)
+		return ret;
 
-	return ret;
-}
+	/* The logical frame size is the width of data for 1 slot */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_I2SCTRL1_REG,
+			~AW88166_I2SFS_MASK, aw88166->fs_value);
+	if (ret)
+		return ret;
 
-static int aw_dev_check_syspll(struct aw_device *aw_dev)
-{
-	int ret;
+	/* The I2S interface mode (Philips standard, LSB/MSB justified) */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_I2SCTRL1_REG,
+			~AW88166_I2SMD_MASK, aw88166->md_value);
+	if (ret)
+		return ret;
 
-	ret = aw_dev_check_mode1_pll(aw_dev);
-	if (ret) {
-		dev_dbg(aw_dev->dev, "mode1 check iis failed try switch to mode2 check");
-		ret = aw_dev_check_mode2_pll(aw_dev);
-		if (ret) {
-			dev_err(aw_dev->dev, "mode2 check iis failed");
-			return ret;
-		}
-	}
+	/* The polarity of the bit clock (BCK) */
+	ret = regmap_update_bits(aw_dev->regmap, AW88166_SYSCTRL_REG,
+			~AW88166_BCKINV_MASK, aw88166->bck_inv_value);
+	if (ret)
+		return ret;
 
-	return 0;
+	return aw_dev_check_pll(aw_dev);
 }
 
 static int aw_dev_check_sysst(struct aw_device *aw_dev)
@@ -732,7 +718,7 @@
 	aw_dev_pwd(aw_dev, false);
 	usleep_range(AW88166_2000_US, AW88166_2000_US + 10);
 
-	ret = aw_dev_check_syspll(aw_dev);
+	ret = aw88166_dev_configure_syspll(aw88166);
 	if (ret) {
 		dev_err(aw_dev->dev, "pll check failed cannot start\n");
 		goto pll_check_fail;
@@ -1254,6 +1240,137 @@
 	return 0;
 }
 
+static int aw88166_set_fmt(struct snd_soc_dai *dai, unsigned int fmt)
+{
+	struct snd_soc_component *component = dai->component;
+	struct aw88166 *aw88166 = snd_soc_component_get_drvdata(component);
+
+	switch (fmt & SND_SOC_DAIFMT_INV_MASK) {
+	case SND_SOC_DAIFMT_NB_NF:
+		aw88166->bck_inv_value = AW88166_BCKINV_NOT_INVERT_VALUE;
+		break;
+	case SND_SOC_DAIFMT_IB_NF:
+		aw88166->bck_inv_value = AW88166_BCKINV_INVERTED_VALUE;
+		break;
+	default:
+		dev_err(aw88166->aw_pa->dev, "unsupported invert mode 0x%x\n",
+			fmt & SND_SOC_DAIFMT_INV_MASK);
+		return -EINVAL;
+	}
+
+	switch (fmt & SND_SOC_DAIFMT_FORMAT_MASK) {
+	case SND_SOC_DAIFMT_I2S:
+		aw88166->md_value = AW88166_I2SMD_PHILIPS_STANDARD_VALUE;
+		break;
+	case SND_SOC_DAIFMT_MSB:
+		aw88166->md_value = AW88166_I2SMD_MSB_JUSTIFIED_VALUE;
+		break;
+	case SND_SOC_DAIFMT_LSB:
+		aw88166->md_value = AW88166_I2SMD_LSB_JUSTIFIED_VALUE;
+		break;
+	default:
+		dev_err(aw88166->aw_pa->dev, "unsupported DAI format 0x%x\n",
+			fmt & SND_SOC_DAIFMT_FORMAT_MASK);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int aw88166_hw_params(struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params,
+	struct snd_soc_dai *dai)
+{
+	struct snd_soc_component *component = dai->component;
+	struct aw88166 *aw88166 = snd_soc_component_get_drvdata(component);
+
+	if (substream->stream == SNDRV_PCM_STREAM_CAPTURE)
+		return 0;
+
+	aw88166->cco_mux_value = AW88166_CCO_MUX_BYPASS_VALUE;
+	switch (params_rate(params)) {
+	case 8000:
+		aw88166->sr_value = AW88166_I2SSR_8KHZ_VALUE;
+		aw88166->cco_mux_value = AW88166_CCO_MUX_DIVIDED_VALUE;
+		break;
+	case 11025:
+		aw88166->sr_value = AW88166_I2SSR_11P025KHZ_VALUE;
+		break;
+	case 12000:
+		aw88166->sr_value = AW88166_I2SSR_12KHZ_VALUE;
+		break;
+	case 16000:
+		aw88166->sr_value = AW88166_I2SSR_16KHZ_VALUE;
+		aw88166->cco_mux_value = AW88166_CCO_MUX_DIVIDED_VALUE;
+		break;
+	case 22050:
+		aw88166->sr_value = AW88166_I2SSR_22P05KHZ_VALUE;
+		break;
+	case 24000:
+		aw88166->sr_value = AW88166_I2SSR_24KHZ_VALUE;
+		break;
+	case 32000:
+		aw88166->sr_value = AW88166_I2SSR_32KHZ_VALUE;
+		aw88166->cco_mux_value = AW88166_CCO_MUX_DIVIDED_VALUE;
+		break;
+	case 44100:
+		aw88166->sr_value = AW88166_I2SSR_44P1KHZ_VALUE;
+		break;
+	case 48000:
+		aw88166->sr_value = AW88166_I2SSR_48KHZ_VALUE;
+		break;
+	case 96000:
+		aw88166->sr_value = AW88166_I2SSR_96KHZ_VALUE;
+		break;
+	default:
+		dev_err(aw88166->aw_pa->dev, "unsupported sample rate %d\n",
+			params_rate(params));
+		return -EINVAL;
+	}
+
+	switch (params_width(params)) {
+	case 16:
+		aw88166->fs_value = AW88166_I2SFS_16_BITS_VALUE;
+		break;
+	case 20:
+		aw88166->fs_value = AW88166_I2SFS_20_BITS_VALUE;
+		break;
+	case 24:
+		aw88166->fs_value = AW88166_I2SFS_24_BITS_VALUE;
+		break;
+	case 32:
+		aw88166->fs_value = AW88166_I2SFS_32_BITS_VALUE;
+		break;
+	default:
+		dev_err(aw88166->aw_pa->dev, "unsupported bit width %d\n",
+			params_width(params));
+		return -EINVAL;
+	}
+
+	switch (params_physical_width(params)) {
+	case 16:
+		aw88166->bck_value = AW88166_I2SBCK_32FS_VALUE;
+		break;
+	case 24:
+		aw88166->bck_value = AW88166_I2SBCK_48FS_VALUE;
+		break;
+	case 32:
+		aw88166->bck_value = AW88166_I2SBCK_64FS_VALUE;
+		break;
+	default:
+		dev_err(aw88166->aw_pa->dev, "unsupported physical bit width %d\n",
+			params_physical_width(params));
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static const struct snd_soc_dai_ops aw88166_dai_ops = {
+	.set_fmt = aw88166_set_fmt,
+	.hw_params = aw88166_hw_params,
+};
+
 static struct snd_soc_dai_driver aw88166_dai[] = {
 	{
 		.name = "aw88166-aif",
@@ -1272,6 +1389,7 @@
 			.rates = AW88166_RATES,
 			.formats = AW88166_FORMATS,
 		},
+		.ops = &aw88166_dai_ops,
 	},
 };
 
@@ -1775,6 +1893,14 @@
 	if (!aw88166)
 		return -ENOMEM;
 
+	/* set defaults */
+	aw88166->sr_value = AW88166_I2SSR_48KHZ_VALUE;
+	aw88166->cco_mux_value = AW88166_CCO_MUX_BYPASS_VALUE;
+	aw88166->fs_value = AW88166_I2SFS_32_BITS_VALUE;
+	aw88166->bck_value = AW88166_I2SBCK_64FS_VALUE;
+	aw88166->bck_inv_value = AW88166_BCKINV_NOT_INVERT_VALUE;
+	aw88166->md_value = AW88166_I2SMD_PHILIPS_STANDARD_VALUE;
+
 	mutex_init(&aw88166->lock);
 
 	i2c_set_clientdata(i2c, aw88166);
--- a/sound/soc/codecs/aw88166.h
+++ b/sound/soc/codecs/aw88166.h
@@ -372,6 +372,110 @@
 		AW88166_PLLS_LOCKED_VALUE | \
 		AW88166_SWS_SWITCHING_VALUE)
 
+#define AW88166_BCKINV_START_BIT	(4)
+#define AW88166_BCKINV_BITS_LEN	(1)
+#define AW88166_BCKINV_MASK	\
+	(~(((1<<AW88166_BCKINV_BITS_LEN)-1) << AW88166_BCKINV_START_BIT))
+
+#define AW88166_BCKINV_NOT_INVERT	(0)
+#define AW88166_BCKINV_NOT_INVERT_VALUE	\
+	(AW88166_BCKINV_NOT_INVERT << AW88166_BCKINV_START_BIT)
+
+#define AW88166_BCKINV_INVERTED	(1)
+#define AW88166_BCKINV_INVERTED_VALUE	\
+	(AW88166_BCKINV_INVERTED << AW88166_BCKINV_START_BIT)
+
+#define AW88166_I2SMD_START_BIT	(8)
+#define AW88166_I2SMD_BITS_LEN	(2)
+#define AW88166_I2SMD_MASK	\
+	(~(((1<<AW88166_I2SMD_BITS_LEN)-1) << AW88166_I2SMD_START_BIT))
+
+#define AW88166_I2SMD_PHILIPS_STANDARD	(0)
+#define AW88166_I2SMD_PHILIPS_STANDARD_VALUE	\
+	(AW88166_I2SMD_PHILIPS_STANDARD << AW88166_I2SMD_START_BIT)
+
+#define AW88166_I2SMD_MSB_JUSTIFIED	(1)
+#define AW88166_I2SMD_MSB_JUSTIFIED_VALUE	\
+	(AW88166_I2SMD_MSB_JUSTIFIED << AW88166_I2SMD_START_BIT)
+
+#define AW88166_I2SMD_LSB_JUSTIFIED	(2)
+#define AW88166_I2SMD_LSB_JUSTIFIED_VALUE	\
+	(AW88166_I2SMD_LSB_JUSTIFIED << AW88166_I2SMD_START_BIT)
+
+#define AW88166_I2SFS_START_BIT		(6)
+#define AW88166_I2SFS_BITS_LEN		(2)
+#define AW88166_I2SFS_MASK		\
+	(~(((1<<AW88166_I2SFS_BITS_LEN)-1) << AW88166_I2SFS_START_BIT))
+
+#define AW88166_I2SFS_16_BITS		(0)
+#define AW88166_I2SFS_16_BITS_VALUE	\
+	(AW88166_I2SFS_16_BITS << AW88166_I2SFS_START_BIT)
+#define AW88166_I2SFS_20_BITS		(1)
+#define AW88166_I2SFS_20_BITS_VALUE	\
+	(AW88166_I2SFS_20_BITS << AW88166_I2SFS_START_BIT)
+#define AW88166_I2SFS_24_BITS		(2)
+#define AW88166_I2SFS_24_BITS_VALUE	\
+	(AW88166_I2SFS_24_BITS << AW88166_I2SFS_START_BIT)
+#define AW88166_I2SFS_32_BITS		(3)
+#define AW88166_I2SFS_32_BITS_VALUE	\
+	(AW88166_I2SFS_32_BITS << AW88166_I2SFS_START_BIT)
+
+#define AW88166_I2SBCK_START_BIT	(4)
+#define AW88166_I2SBCK_BITS_LEN	(2)
+#define AW88166_I2SBCK_MASK	\
+	(~(((1<<AW88166_I2SBCK_BITS_LEN)-1) << AW88166_I2SBCK_START_BIT))
+
+#define AW88166_I2SBCK_32FS	(0)
+#define AW88166_I2SBCK_32FS_VALUE	\
+	(AW88166_I2SBCK_32FS << AW88166_I2SBCK_START_BIT)
+
+#define AW88166_I2SBCK_48FS	(1)
+#define AW88166_I2SBCK_48FS_VALUE	\
+	(AW88166_I2SBCK_48FS << AW88166_I2SBCK_START_BIT)
+
+#define AW88166_I2SBCK_64FS	(2)
+#define AW88166_I2SBCK_64FS_VALUE	\
+	(AW88166_I2SBCK_64FS << AW88166_I2SBCK_START_BIT)
+
+#define AW88166_I2SSR_START_BIT		(0)
+#define AW88166_I2SSR_BITS_LEN		(4)
+#define AW88166_I2SSR_MASK		\
+	(~(((1<<AW88166_I2SSR_BITS_LEN)-1) << AW88166_I2SSR_START_BIT))
+
+#define AW88166_I2SSR_8KHZ		(0)
+#define AW88166_I2SSR_8KHZ_VALUE	\
+	(AW88166_I2SSR_8KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_11P025KHZ		(1)
+#define AW88166_I2SSR_11P025KHZ_VALUE	\
+	(AW88166_I2SSR_11P025KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_12KHZ		(2)
+#define AW88166_I2SSR_12KHZ_VALUE	\
+	(AW88166_I2SSR_12KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_16KHZ		(3)
+#define AW88166_I2SSR_16KHZ_VALUE	\
+	(AW88166_I2SSR_16KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_22P05KHZ		(4)
+#define AW88166_I2SSR_22P05KHZ_VALUE	\
+	(AW88166_I2SSR_22P05KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_24KHZ		(5)
+#define AW88166_I2SSR_24KHZ_VALUE	\
+	(AW88166_I2SSR_24KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_32KHZ		(6)
+#define AW88166_I2SSR_32KHZ_VALUE	\
+	(AW88166_I2SSR_32KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_44P1KHZ		(7)
+#define AW88166_I2SSR_44P1KHZ_VALUE	\
+	(AW88166_I2SSR_44P1KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_48KHZ		(8)
+#define AW88166_I2SSR_48KHZ_VALUE	\
+	(AW88166_I2SSR_48KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_96KHZ		(9)
+#define AW88166_I2SSR_96KHZ_VALUE	\
+	(AW88166_I2SSR_96KHZ << AW88166_I2SSR_START_BIT)
+#define AW88166_I2SSR_192KHZ		(10)
+#define AW88166_I2SSR_192KHZ_VALUE	\
+	(AW88166_I2SSR_192KHZ << AW88166_I2SSR_START_BIT)
+
 #define AW88166_CCO_MUX_START_BIT	(14)
 #define AW88166_CCO_MUX_BITS_LEN	(1)
 #define AW88166_CCO_MUX_MASK		\
@@ -459,7 +563,10 @@
 #define AW88166_I2C_NAME			"aw88166"
 #define AW88166_ACF_FILE			"aw88166_acf.bin"
 
+/* NOTE: 192000 has a reg value downstream but is not advertised here */
 #define AW88166_RATES (SNDRV_PCM_RATE_8000_48000 | \
+			SNDRV_PCM_RATE_12000 | \
+			SNDRV_PCM_RATE_24000 | \
 			SNDRV_PCM_RATE_96000)
 #define AW88166_FORMATS (SNDRV_PCM_FMTBIT_S16_LE | \
 			SNDRV_PCM_FMTBIT_S24_LE | \
